### ISSUE TITLE
Not possible to modify the options of the razor view engine

### ISSUE THREAD

**Issue by simonech** [umbraco/Umbraco-CMS#12157]
### Which *exact* Umbraco version are you using? For example: 9.0.1 - don't just write v9

9.3.0

### Bug summary

I've been trying to implement support for "themes" in my view rendering and after a bit of digging in the code I realized Umbraco is "wrapping" the default razor view engine into a few additional wrapping engines, which makes it impossible to set options.

### Specifics

On a normal ASP.NET Core application, one would just set new values for `RazorViewEngineOptions.ViewLocationFormats` or by creating a new `ViewLocationExpander` class. 

I've been trying to do it with a `Composer` but any change done there never get executed. The only way to inject new view location is by adding directly a new builder method to be called in the startup class

```cs
            services.AddUmbraco(_env, _config)
                .AddBackOffice()
                .AddThemes() //custom extension that adds a new ViewLocationExpander similar to the core umbraco one
                .AddWebsite()
                .AddComposers()
                .Build();
```
This adds the search path, but at the bottom of the search list, so not useful for our context.

A possible cause could be the[`RefreshingRazorViewEngine`](https://github.com/umbraco/Umbraco-CMS/blob/v9/contrib/src/Umbraco.Web.Common/ModelsBuilder/RefreshingRazorViewEngine.cs) that contains a copy of the razor view engine, so setting the options once this has been wrapped has no effect on the view locations.

Another possible cause of the problem could come from the [`ProfilingViewEngine`](https://github.com/umbraco/Umbraco-CMS/blob/v9/contrib/src/Umbraco.Web.Website/ViewEngines/ProfilingViewEngine.cs) which also wraps the current view engine.

Frankly I've no idea how this could be solved, but this makes it impossible to change view locations, and probably also to replace the view engine with something other than Razor (since ModelsBuilder requires it)

### Steps to reproduce

Here is the sample Composer

```cs
    public class ThemedViewComposer : IComposer
    {
        public void Compose(IUmbracoBuilder builder)
        {
            builder.Services.ConfigureOptions<ThemedViewEngineOptionsSetup>();
        }
    }
```

And here the Options setup class

```cs
    public class ThemedViewEngineOptionsSetup : IConfigureOptions<RazorViewEngineOptions>
    {
        public void Configure(RazorViewEngineOptions options)
        {
            if (options == null)
            {
                throw new ArgumentNullException(nameof(options));
            }

            options.ViewLocationExpanders.Add(new ViewLocationExpander());
        }
    }

    internal class ViewLocationExpander : IViewLocationExpander
    {
        public IEnumerable<string> ExpandViewLocations(ViewLocationExpanderContext context, IEnumerable<string> viewLocations)
        {

            string[] themedLocations = new string[]
            {
                "/Views/mytheme/{0}.cshtml",
                "/Views/mytheme/Shared/{0}.cshtml",
                "/Views/mytheme/Partials/{0}.cshtml",
                "/Views/mytheme/MacroPartials/{0}.cshtml",
            };

            viewLocations = themedLocations.Concat(viewLocations);

            return viewLocations;
        }

        // not a dynamic expander
        public void PopulateValues(ViewLocationExpanderContext context) { }
    }
```

### Expected result / actual result

I'd like to be able to write a Composer that adds viewlocations or ViewLocationExpanders or even adds a new view engine replacing Razor.

**Comment by PhyxionNL** [comment]
I ran into the same problem. It's not because of the profiling view engine but because RefreshingRazorViewEngine is coping its services (not sure if that's correct?), which makes it impossible to modify afterwards. After some trial and error, it is possible to change. You need to define a custom razor view engine that changes the ViewLocationExpanders order, see below:

```csharp
public class PrioritizingRazorViewEngine : IRazorViewEngine
{
    private readonly RazorViewEngine _razorViewEngine;

    /// <summary>
    /// Initializes a new instance of the <see cref="PrioritizingRazorViewEngine"/> class.
    /// </summary>
    /// <param name="pageFactory">The page factory.</param>
    /// <param name="pageActivator">The page activator.</param>
    /// <param name="htmlEncoder">The HTML encoder.</param>
    /// <param name="optionsAccessor">The options accessor.</param>
    /// <param name="loggerFactory">The logger factory.</param>
    /// <param name="diagnosticListener">The diagnostic listener.</param>
    public PrioritizingRazorViewEngine
    (
        IRazorPageFactoryProvider pageFactory,
        IRazorPageActivator pageActivator,
        HtmlEncoder htmlEncoder,
        IOptions<RazorViewEngineOptions> optionsAccessor,
        ILoggerFactory loggerFactory,
        DiagnosticListener diagnosticListener)
    {
        var razorViewEngineOptions = optionsAccessor.Value;

        // Your own view location expander will be the first item, this needs to be at the end to override the searched locations.
        var ownViewLocationExpander = razorViewEngineOptions.ViewLocationExpanders[0];
        razorViewEngineOptions.ViewLocationExpanders.RemoveAt(0);
        razorViewEngineOptions.ViewLocationExpanders.Add(ownViewLocationExpander);

        _razorViewEngine = new RazorViewEngine(pageFactory, pageActivator, htmlEncoder, optionsAccessor, loggerFactory, diagnosticListener);
    }

    /// <inheritdoc />
    public ViewEngineResult FindView(ActionContext context, string viewName, bool isMainPage)
    {
        return _razorViewEngine.FindView(context, viewName, isMainPage);
    }

    /// <inheritdoc />
    public ViewEngineResult GetView(string executingFilePath, string viewPath, bool isMainPage)
    {
        return _razorViewEngine.GetView(executingFilePath, viewPath, isMainPage);
    }

    /// <inheritdoc />
    public RazorPageResult FindPage(ActionContext context, string pageName)
    {
        return _razorViewEngine.FindPage(context, pageName);
    }

    /// <inheritdoc />
    public RazorPageResult GetPage(string executingFilePath, string pagePath)
    {
        return _razorViewEngine.GetPage(executingFilePath, pagePath);
    }

    /// <inheritdoc />
    public string GetAbsolutePath(string executingFilePath, string pagePath)
    {
        return _razorViewEngine.GetAbsolutePath(executingFilePath, pagePath);
    }
}
```

Then **before** services.AddUmbraco:

```csharp
    services.ConfigureOptions<ThemedViewEngineOptionsSetup>(); // IConfigureOptions<RazorViewEngineOptions>.
    services.AddSingleton<IRazorViewEngine, PrioritizingRazorViewEngine>();
```

You don't need the composer.  The RefreshingRazorViewEngine is still used as it now wraps around the PrioritizingRazorViewEngine, although I don't think you need this if you build models to source code instead of in memory.
Basically it'll now wrap as: ProfilingRazorViewEngine > RefreshingRazorViewEngine > PrioritizingRazorViewEngine > RazorViewEngine.

**Comment by simonech** [comment]
Indeed, it's the `RefreshingRazorViewEngine` that causes the problem.
I guess one possible "bug" discovered by this issue is that the startup process should not use the RefreshingRazorViewEngine when not needed for the InMemory model builder.

Thanks for the tip!

**Comment by kjac** [comment]
Hi @simonech,

Thanks for reaching out. Sorry for the belated response. I hope you managed to apply the workaround from @PhyxionNL or something similar in the mean time. And thank you @PhyxionNL for supplying that 💪 

You can indeed add additional view locations without replacing the entire `IRazorViewEngine`, but it is a little bit tricky. You'll need this code:

```csharp
public class CustomViewEngineOptionsSetup : IConfigureOptions<RazorViewEngineOptions>
{
    public void Configure(RazorViewEngineOptions options)
    {
        if (options == null)
        {
            throw new ArgumentNullException(nameof(options));
        }

        options.ViewLocationExpanders.Add(new CustomViewLocationExpander());
    }

    private class CustomViewLocationExpander : IViewLocationExpander
    {
        public IEnumerable<string> ExpandViewLocations(ViewLocationExpanderContext context, IEnumerable<string> viewLocations)
        {
            // list of additional view locations to search when resolving a view
            string[] themedLocations = new string[]
            {
                "/OtherViews/{0}.cshtml"
            };

            viewLocations = themedLocations.Concat(viewLocations);

            return viewLocations;
        }

        public void PopulateValues(ViewLocationExpanderContext context) { }
    }
}

public static class CustomUmbracoBuilderExtensions
{
    public static IUmbracoBuilder AddCustomViewLocations(this IUmbracoBuilder builder)
    {
        builder.Services.ConfigureOptions<CustomViewEngineOptionsSetup>();
        return builder;
    }
}
```

Next you need to apply the custom view locations *before* the frontend website is initialized:

```csharp
public void ConfigureServices(IServiceCollection services)
{
    services.AddUmbraco(_env, _config)
        .AddBackOffice()
        // add custom view locations before AddWebsite is called
        .AddCustomViewLocations()
        .AddWebsite()
        .AddComposers()
        .Build();
}
```

Now the Razor view engine will search for views in all the Umbraco default locations *and* your custom locations. For example:

![image](https://user-images.githubusercontent.com/7405322/211488019-b09f18be-e189-4685-98e0-f90dbc6bd03d.png)

...where `Home.cshtml` is:

```cshtml
@using Umbraco.Cms.Web.Common.PublishedModels;
@inherits Umbraco.Cms.Web.Common.Views.UmbracoViewPage<ContentModels.Home>
@using ContentModels = Umbraco.Cms.Web.Common.PublishedModels;
@{
	Layout = null;
}
<body>
<h1>@Model.Title</h1>
@Html.Partial("MyPartialView")
@Html.Partial("MyOtherPartialView")
</body>
```

...yields:

![image](https://user-images.githubusercontent.com/7405322/211488242-26e803d8-7b29-4dc5-9800-7fed64c7db55.png)

**Comment by enkelmedia** [comment]
Just wanted to add to this that the example provided by @kjac does not work on v13.5.2.

I tried the exact steps but the views in my "themes"-folder did not resolve. After adding the `PrioritizingRazorViewEngine` as outlined by @PhyxionNL the view was resolved on the first try.

Clearly @kjac got it to work in the example but it looks like something is missing from the steps provided.

**Comment by umbracocommunity** [comment]
This issue has been mentioned on **Umbraco community forum**. There might be relevant details there:

https://forum.umbraco.com/t/how-to-handle-template-file-save-location-in-umbraco-13-multi-site-project-with-usync/3455/5

**Comment by umbracocommunity** [comment]
This issue has been mentioned on **Umbraco community forum**. There might be relevant details there:

https://forum.umbraco.com/t/razor-class-library-and-umbraco-with-runtime-compilation/3488/9
